zyte_common_items/components.py

```diff
@@ -108,7 +109,7 @@ class Image(Item):
     #: `Data URIs`_ are not allowed in this attribute.
     #:
     #: .. _Data URIs: https://en.wikipedia.org/wiki/Data_URI_scheme
-    url: str
+    url: str = attrs.field(converter=str)
```

Comment:
I wonder if we should support only Url classes from web-poet, not arbitrary objects. It seems with the current implementation you can pass anything to url (e.g. a Selector instance, or a list of strings), and it won't raise a type error.

As discussed elsewhere, it'd also be good to ensure conversion happens on assignment.

Reply:
> it'd also be good to ensure conversion happens on assignment.
Added tests cases in https://github.com/zytedata/zyte-common-items/pull/11/commits/7e9135dbf893921012b9de4d7325e1b33fc3de33.

> I wonder if we should support only Url classes from web-poet, not arbitrary objects. It seems with the current implementation you can pass anything to url (e.g. a Selector instance, or a list of strings), and it won't raise a type error.
I've partially implemented this locally but need to add more test cases to handle some use cases.

Reply:
So I've pushed some changes in https://github.com/zytedata/zyte-common-items/pull/11/commits/3e53d6154ce7e5f1f9d0ec85f7768c100de3ad7f which:

- ensures that values that are not URL classes (i.e. subclasses of `web_poet.page_inputs.url._Url`) are not auto-converted. For example, `Image(url=123)` should become `Image(url="123")`.
- ensure that **mypy** raises errors when assigning non-string or non-URL values to fields. For example, doing `img.url = 123` should error out in **mypy**.

However, one thing's that lacking is ensuring that **mypy** also raises the appropriate errors when instantiating with incorrect types. For example, `Image(url=123)` should cause **mypy** raise an error. For now, I've marked the tests with `xfail` ([reference](https://github.com/zytedata/zyte-common-items/pull/11/commits/3e53d6154ce7e5f1f9d0ec85f7768c100de3ad7f#diff-901c8a5cbbdeeb7f1d4acf5ed1da5352f3f6fa461e1f1d11de2bf8a02d5f9c08R121-R164)) until **web-poet** uploads its stubs as [mentioned here](https://github.com/zytedata/zyte-common-items/pull/11/commits/3e53d6154ce7e5f1f9d0ec85f7768c100de3ad7f#diff-901c8a5cbbdeeb7f1d4acf5ed1da5352f3f6fa461e1f1d11de2bf8a02d5f9c08R107-R118).